browser/net/ipfs_redirect_network_delegate_helper.cc

```diff
@@ -7,15 +7,45 @@
 
 #include <string>
 
+#include "brave/browser/ipfs/ipfs_service_factory.h"
 #include "brave/browser/profiles/profile_util.h"
+#include "brave/components/ipfs/ipfs_service.h"
 #include "brave/components/ipfs/ipfs_utils.h"
 #include "chrome/common/channel_info.h"
 #include "components/user_prefs/user_prefs.h"
 #include "content/public/browser/browser_context.h"
 #include "net/base/net_errors.h"
+#include "net/base/url_util.h"
+#include "url/url_constants.h"
 
 namespace ipfs {
 
+int OnBeforeStartTransaction_AddXForwardedProtoHeader(
+    net::HttpRequestHeaders* headers,
+    const brave::ResponseCallback& next_callback,
+    std::shared_ptr<brave::BraveRequestInfo> ctx) {
+  // Add header when navigate to localhost/ipns/<path> so local node could
+  // redirect to proper url under ipns.localhost domain.
+  if (ctx->request_url.SchemeIsHTTPOrHTTPS() &&
+      net::IsLocalhost(ctx->request_url) && HasIPNSPath(ctx->request_url)) {
+    headers->AddHeadersFromString("X-Forwarded-Proto: https");
+  }
+  return net::OK;
+}
+
+int OnBeforeURLRequest_LocalIPFSNodeSchemeManager(
+    const brave::ResponseCallback& next_callback,
+    std::shared_ptr<brave::BraveRequestInfo> ctx) {
+  // Reduce HTTPS to HTTP because local IPFS node can't handle HTTPS.
+  if (ctx->request_url.DomainIs("ipns.localhost") &&
```

Comment:
is this right? Can't the domain be a.b.ipns.localhost at this point?

Reply:
>   // Returns true if this URL's host matches or is in the same domain as
>   // the given input string. For example, if the hostname of the URL is
>   // "www.google.com", this will return true for "com", "google.com", and
>   // "www.google.com".
>

Reply:
yes, but it doesn't get the etld, it just separates hostname from the rest. You need to get use net::registry_controlled_domains::GetDomainAndRegistry here

Reply:
also there are no unit tests for this method to verify the functionality

Reply:
`localhost` is not on the public suffix list anyway

Reply:
@iefremov `ipns.localhost` is and that's what matter. This won't work correctly if you have a.b.ipns.localhost because the domain will be b.ipns.localhost and won't match

Reply:
cc @pes10k

Reply:
so actually `DomainIs` is fine. It's been a while and I got things reversed in my head, but adding `X-Forwarded-Proto: https` converts dot to dash so a.b becomes a-b.ipns.localhost and then DomainIs works fine

Reply:
GURL::DomainIs just separates host by "." and tests equivalence for a.b.ipns.localhost b.ipns.localhost ipns.localhost. It doesn't look to public suffix list.

Reply:
just for posterity, this is ok because it maps `.` to `-` for the hostname anyway